impl_core/src/main/java/io/opencensus/implcore/trace/propagation/TraceContextFormat.java

```diff
@@ -0,0 +1,170 @@
+    return FIELDS;
+  }
+
+  @Override
+  public <C /*>>> extends @NonNull Object*/> void inject(
+      SpanContext spanContext, C carrier, Setter<C> setter) {
+    checkNotNull(spanContext, "spanContext");
+    checkNotNull(setter, "setter");
+    checkNotNull(carrier, "carrier");
+    setter.put(
+        carrier,
+        TRACEPARENT,
+        TRACEPARENT_HEADER_PREFIX
+            + spanContext.getTraceId().toLowerBase16()
+            + TRACEPARENT_DELIMITER
+            + spanContext.getSpanId().toLowerBase16()
+            + TRACEPARENT_DELIMITER
+            + spanContext.getTraceOptions().toLowerBase16());
+    List<Tracestate.Entry> entries = spanContext.getTracestate().getEntries();
+    if (entries.isEmpty()) {
+      // No need to add an empty "tracestate" header.
+      return;
+    }
+    StringBuilder stringBuilder = new StringBuilder(TRACESTATE_MAX_SIZE);
+    for (Tracestate.Entry entry : entries) {
+      if (stringBuilder.length() != 0) {
+        stringBuilder.append(TRACESTATE_ENTRY_DELIMITER);
+      }
+      stringBuilder
+          .append(entry.getKey())
+          .append(TRACESTATE_KEY_VALUE_DELIMITER)
+          .append(entry.getValue());
+    }
+    setter.put(carrier, TRACESTATE, stringBuilder.toString());
+  }
+
+  @Override
+  public <C /*>>> extends @NonNull Object*/> SpanContext extract(C carrier, Getter<C> getter)
+      throws SpanContextParseException {
+    checkNotNull(carrier, "carrier");
+    checkNotNull(getter, "getter");
+    TraceId traceId;
+    SpanId spanId;
+    TraceOptions traceOptions;
+    String traceparent = getter.get(carrier, TRACEPARENT);
+    try {
+      // TODO(bdrutu): Do we need to verify that version is hex and that for the version
+      // the length is the expected one?
+      checkArgument(
+          traceparent != null
+              && traceparent.charAt(TRACE_OPTION_OFFSET - 1) == TRACEPARENT_DELIMITER
+              && (traceparent.length() == TRACEPARENT_HEADER_SIZE
+                  || (traceparent.length() > TRACEPARENT_HEADER_SIZE
+                      && traceparent.charAt(TRACEPARENT_HEADER_SIZE) == TRACEPARENT_DELIMITER))
+              && traceparent.charAt(SPAN_ID_OFFSET - 1) == TRACEPARENT_DELIMITER
+              && traceparent.charAt(TRACE_OPTION_OFFSET - 1) == TRACEPARENT_DELIMITER,
+          "Missing or malformed TRACEPARENT.");
+
+      traceId =
+          TraceId.fromLowerBase16(
+              CharBuffer.wrap(traceparent, TRACE_ID_OFFSET, TRACE_ID_OFFSET + TRACE_ID_HEX_SIZE));
+      spanId =
+          SpanId.fromLowerBase16(
+              CharBuffer.wrap(traceparent, SPAN_ID_OFFSET, SPAN_ID_OFFSET + SPAN_ID_HEX_SIZE));
+      traceOptions =
+          TraceOptions.fromLowerBase16(
+              CharBuffer.wrap(
+                  traceparent, TRACE_OPTION_OFFSET, TRACE_OPTION_OFFSET + TRACE_OPTION_HEX_SIZE));
+    } catch (IllegalArgumentException e) {
+      throw new SpanContextParseException("Invalid traceparent: " + traceparent, e);
```

Comment:
I think failure to parse should be more lenient with this spec. Even better if the constructor *requires* passing a fallback format for now. In brave, I will suggest:

* attempt to parse traceparent, but ignore some parsing due to thrash on flags. I know here you aren't interpreting flags directly, but anyway this part is a bit useless to us now
* attempt to parse tracestate (with a format configured, ex a map of "b3" to B3SingleFormat). If that passes good.
* if this also crashes, parse the fallback format

IOTW we don't ever crash on an unstable format. thoughts?